File: 1872-stone-game-viii/1872-stone-game-viii.cpp
class Solution {
public:
    int dp[100001]= {[0 ... 100000]=INT_MIN};
    int recur(int i,vector<int>&stones){
        if(i==stones.size()-1)return stones[i];
        if(dp[i]!=INT_MIN)return dp[i];
        return dp[i]= max(recur(i+1,stones),stones[i]-recur(i+1,stones));
    }
    
    int stoneGameVIII(vector<int>& stones) {
        partial_sum(stones.begin(),stones.end(),stones.begin());
        return recur(1,stones);
        
    }
};

File: 1872-stone-game-viii/1872-stone-game-viii.java
class Solution {
    
    int []dp;
    // Top DOWN approach
    int memo(int []stones,int idx)
    {
        if(idx==stones.length-1)return stones[idx];
        if(dp[idx]!=Integer.MIN_VALUE)return dp[idx];
        return dp[idx]= Math.max(memo(stones,idx+1),stones[idx]-memo(stones,idx+1));
    }
    
    public int stoneGameVIII(int[] stones) {
        int n=stones.length;
        for(int i=1;i<n;i++)stones[i]+=stones[i-1];
        dp=new int[n];
        // Arrays.fill(dp,Integer.MIN_VALUE);
        // return memo(stones,1);
        
        //Bottom UP Approach
        dp[n-1]=stones[n-1];
        for(int i=n-2;i>0;i--){
            dp[i]=Math.max(stones[i]-dp[i+1],dp[i+1]);
        }
        return dp[1];
    }
}